mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/impl/MongoDBIndexesDefiner.java

```diff
@@ -147,11 +147,36 @@ void validateIndexSpecsForIndexes(Table table, IndexOptions indexOptions, Set<St
 			MongoDBIndexSpec indexSpec = new MongoDBIndexSpec( index,
 					getIndexOptionDocument( table, indexOptions.getOptionForIndex( index.getName() ) ) );
 			if ( validateIndexSpec( indexSpec ) ) {
-				indexSpecs.add( indexSpec );
+				addHandleDuplications( indexSpec );
 			}
 		}
 	}
 
+	private void addHandleDuplications(MongoDBIndexSpec indexSpec) {
+		// we can add any index defined on different collection or different key set
+		// we rely on hashcode and equals to verify conflicts
+		if ( !indexSpecs.contains( indexSpec ) ) {
+			indexSpecs.add( indexSpec );
+			return;
+		}
+
+		int existingIndexPosition = indexSpecs.indexOf( indexSpec );
+		MongoDBIndexSpec existingIndex = indexSpecs.get( existingIndexPosition );
+
+		// in case on conflict
+		// we rely on compareTo to choose the best candidate
+		if ( indexSpec.compareTo( existingIndex ) < 0 ) {
+			// replace the existing definition with the new one
+			indexSpecs.remove( existingIndexPosition );
+			indexSpecs.add( indexSpec );
+			log.tryToDefineMultipleIndexesOnTheSameKeySet( indexSpec.getIndexKeysDocument().toJson(), existingIndex.getIndexName() );
+		}
+		else {
+			// discard the current definition
+			log.tryToDefineMultipleIndexesOnTheSameKeySet( indexSpec.getIndexKeysDocument().toJson(), indexSpec.getIndexName() );
```

Comment:
I don't understand what's going on here. How the compare to helps you to choose the right candidate?

Reply:
Also, why you log the same message in both cases?

Reply:
Here I reach the else branch if have 2 indexes, both defined on the same field set of the same collection. The solution proposed is to skip the definition of one of them.

Reply:
The role of *compareTo* here is to choose one of them. The choose should be repeatable with the same result, so we need a fixed criteria. BTW we can avoid to use *compareTo* to this.

Reply:
the `log` statement will need guarding. Please check if the cathegory of the log level is high enough to not perform a `toJson()` operation for no reason.

Reply:
Severity level is warning here, the row is executed every time a user tries to define another index on the same fields of a pre-existing one.

Reply:
Then the operation is performed rarely. I would prefer leave `toJSon`. But as you prefer, it is not critical to me.

Reply:
I think the point is to add a check like `if ( log.isEnabled( Logger.Level.DEBUG.WARN) ) { ... }`.
You don't have to remove the call to toJSon().

Reply:
Does it make sense guarding a WARN? It's just a curiosity. Anyway I'm going to to that.

Reply:
I guess in production somebody might be only interested to see ERRORs